src/nxt_array.c

```diff
@@ -1,150 +1,253 @@
 
 /*
+ * Copyright (C) Evgenii Sokolov
  * Copyright (C) Igor Sysoev
  * Copyright (C) NGINX, Inc.
  */
 
 #include <nxt_main.h>
 
```

Comment:
It seems you broke the design, I don't understand why you thought there is a problem with the current way.

Take a look at the change.

```
[](https://github.com/nginx/unit/pull/664/files#diff-b340fa7a4c4fbec4cfadd48c7469e88ccb7a183b31423c74a0d5a23047d11e85)[](https://github.com/nginx/unit/pull/664/files#diff-b340fa7a4c4fbec4cfadd48c7469e88ccb7a183b31423c74a0d5a23047d11e85)struct nxt_var_query_s {
    nxt_array_t         values;   /* of nxt_var_value_t */
    nxt_array_t         parts;    /* of nxt_str_t * */
    nxt_array_t         *values;   /* of nxt_var_value_t */
    nxt_array_t         *parts;    /* of nxt_str_t * */

    nxt_lvlhsh_t        cache;

@@ -433,11 +433,18 @@ nxt_var_query_init(nxt_var_query_t **query_p, void *ctx, nxt_mp_t *mp)
            return NXT_ERROR;
        }

        nxt_array_init(&query->values, mp, sizeof(nxt_var_value_t));
        nxt_array_init(&query->parts, mp, sizeof(nxt_str_t *));
        query->values = nxt_array_create(mp, 1, sizeof(nxt_var_value_t));
        if (nxt_slow_path(query->values == NULL)) {
            return NXT_ERROR;
        }

        query->parts = nxt_array_create(mp, 1, sizeof(nxt_str_t *));
        if (nxt_slow_path(query->parts == NULL)) {
            return NXT_ERROR;
        }
```

That's a few more allocations than before, and that's why `nxt_array_init()` is required.

Reply:
Hello @hongzhidao, 
Thank you for your attention to my suggested fixes and improvements!

I want to draw your attention to the fact that `nxt_array_init` has been completely removed and replaced in favor of `nxt_array_create`.

I thank you for bringing my additional attention to the concepts used in **nxt_var.c**. In this regard, I paid more attention to the proposed changes and related objects.

I came to the conclusion that in the next commit to the PR I will propose changes taking into account the need to use `nxt_array_init` and not changing **nxt_var.c** due to the peculiarities of working with the `nxt_var_query_s` structure, since some of its properties are closely related to caching for which I would not want influence in any way.

I'll explain why I'm proposing these changes in my next answer. 
Thank you for your attention!